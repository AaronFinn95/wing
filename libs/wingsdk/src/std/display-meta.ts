/**
 * Display meta component.
 */
export class DisplayMetaComponent {
  constructor(
    /**
     * The component type.
     */
    public readonly type: string,
    /**
     * The component props.
     */
    public readonly props: { [key: string]: any }
  ) {}
}

/**
 * Display meta link props
 */
export interface LinkProps {
  /**
   * The link href.
   */
  readonly href: string;
  /**
   * The link text.
   */
  readonly text: string;
}

/**
 * Display meta link component.
 */
export class DisplayMetaLink extends DisplayMetaComponent {
  constructor(props: LinkProps) {
    super("link", props);
  }
}

/**
 * Display meta
 */
export class DisplayMeta {
  private readonly _components: DisplayMetaComponent[];
  constructor() {
    this._components = [];
  }

  /**
   * Returns the meta components.
   */
  public get components(): DisplayMetaComponent[] | undefined {
    return [...this._components];
  }

  private addComponent(component: DisplayMetaComponent) {
    this._components.push(component);
  }

  /**
   * Add a link to the array of meta components.
   * @param href
   * @param text
   */
  public addLink(href: string, text: string) {
    this.addComponent(new DisplayMetaLink({ href, text }));
  }
}
