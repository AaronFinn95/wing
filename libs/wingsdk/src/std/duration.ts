/**
 * Represents a length of time.
 */
export class Duration {
  /**
   * Create a Duration representing an amount of minutes
   *
   * @param amount the amount of Minutes the `Duration` will represent.
   * @returns a new `Duration` representing `amount` Minutes.
   */
  public static fromMinutes(amount: number): Duration {
    return new Duration(amount * 60 * 1000);
  }

  /**
   * Create a Duration representing an amount of hours
   *
   * @param amount the amount of Hours the `Duration` will represent.
   * @returns a new `Duration` representing `amount` Hours.
   */
  public static fromHours(amount: number): Duration {
    return new Duration(amount * 60 * 60 * 1000);
  }

  /**
   * Create a Duration representing an amount of seconds
   *
   * @param amount the amount of Seconds the `Duration` will represent.
   * @returns a new `Duration` representing `amount` Seconds.
   */
  public static fromSeconds(amount: number): Duration {
    return new Duration(amount * 1000);
  }

  /**
   * Create a Duration representing an amount of milliseconds
   *
   * @param amount the amount of milliseconds the `Duration` will represent.
   * @returns a new `Duration` representing `amount` milliseconds.
   */
  public static fromMilliseconds(amount: number): Duration {
    return new Duration(amount);
  }

  /**
   * Return the total number of milliseconds in this Duration
   *
   * @returns the value of this `Duration` expressed in Milliseconds.
   */
  public readonly milliseconds: number;

  private constructor(milliseconds: number) {
    this.milliseconds = milliseconds;
  }

  /**
   * Return the total number of minutes in this Duration
   *
   * @returns the value of this `Duration` expressed in Minutes.
   */
  public get minutes() {
    return this.seconds / 60;
  }

  /**
   * Return the total number of hours in this Duration
   *
   * @returns the value of this `Duration` expressed in Hours.
   */
  public get hours() {
    return this.minutes / 60;
  }

  /**
   * Return the total number of milliseconds in this Duration
   *
   * @returns the value of this `Duration` expressed in Milliseconds.
   */
  public get seconds() {
    return this.milliseconds / 1000;
  }
}
