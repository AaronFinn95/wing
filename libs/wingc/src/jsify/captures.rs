use crate::{
	ast::{
		ArgList, Class, Expr, ExprKind, FunctionBody, Phase, Reference, Scope, Stmt, Symbol, TypeAnnotation,
		TypeAnnotationKind,
	},
	diagnostic::{report_diagnostic, Diagnostic, WingSpan},
	type_check::{
		resolve_user_defined_type,
		symbol_env::{LookupResult, SymbolEnv, SymbolEnvRef},
		Type, TypeRef, Types, UnsafeRef, VariableInfo,
	},
	visit::{self, Visit},
};
use indexmap::IndexMap;
use itertools::Itertools;
use std::{
	cell::RefCell,
	collections::{BTreeMap, BTreeSet},
	fmt::Display,
	vec,
};

/// Lists all the captures from a class
pub struct ClassCaptures {
	// all the types captured by this class
	types: IndexMap<Vec<String>, TypeRef>,

	/// captured variables per method
	vars: BTreeMap<String, BTreeMap<String, BTreeSet<String>>>,
}

/// Visitor that finds all the types and variables used within an inflight method but defined in its
/// parent environment
pub struct CaptureScanner<'a> {
	/// The name of the method
	method_name: &'a str,

	/// The root scope of the function we're analyzing
	function_scope: &'a Scope,

	/// The method env, used to lookup the type
	method_env: &'a RefCell<Option<SymbolEnv>>,

	/// The current environment (tracked via the visitor)
	current_env: &'a RefCell<Option<SymbolEnv>>,

	/// The index of the last visited statement.
	current_index: usize,

	/// The types table
	types: &'a Types,

	captures: &'a mut ClassCaptures,
}

#[derive(Clone)]
struct Component {
	text: String,
	span: WingSpan,
	kind: ComponentKind,
}

#[derive(Clone)]
enum ComponentKind {
	Member(VariableInfo, SymbolEnvRef),
	ClassType(TypeRef),
}

impl Display for Component {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		write!(f, "{}", self.text)
	}
}

impl ClassCaptures {
	/// Scan all inflight methods in a class and extract the names of all the types and variables that
	/// are defined outside of this method.
	pub fn scan(types: &Types, class: &Class) -> ClassCaptures {
		let mut captures = ClassCaptures::new();

		for (method_name, m) in &class.methods {
			// only consider inflight methods
			if m.signature.phase != Phase::Inflight {
				continue;
			}

			if let FunctionBody::Statements(scope) = &m.body {
				let mut visitor = CaptureScanner::new(types, &method_name.name, scope, &mut captures);
				visitor.scan();
			}
		}

		// Remove myself from the list of referenced preflight types because I don't need to import myself
		captures.remove_type(&vec![class.name.name.clone()]);

		captures
	}

	fn new() -> Self {
		Self {
			types: IndexMap::new(),
			vars: BTreeMap::new(),
		}
	}

	pub fn capture_type(&mut self, name: Vec<String>, t: UnsafeRef<Type>) {
		self.types.insert(name, t);
	}

	pub fn capture_var(&mut self, method_name: &str, var: &str, op: Option<String>) {
		let entry = self
			.vars
			.entry(method_name.to_string())
			.or_default()
			.entry(var.to_string())
			.or_default();

		if let Some(op) = op {
			entry.insert(op);
		}
	}

	pub fn remove_type(&mut self, name: &[String]) {
		self.types.remove(name);
	}

	pub fn vars(&self) -> IndexMap<String, BTreeSet<String>> {
		let mut result: IndexMap<String, BTreeSet<String>> = IndexMap::new();

		for (_, captures) in &self.vars {
			for (var, ops) in captures {
				result.entry(var.to_string()).or_default().extend(ops.clone());
			}
		}

		result
	}

	pub fn free_vars_and_types(&self) -> BTreeSet<String> {
		let mut result = BTreeSet::new();

		for (v, _) in self.vars() {
			if !v.starts_with("this.") {
				result.insert(v);
			}
		}

		for (t, _) in self.types() {
			result.insert(t.join("_"));
		}

		return result;
	}

	pub fn free_vars(&self) -> BTreeMap<String, BTreeSet<String>> {
		self
			.vars()
			.iter()
			.filter(|(v, _)| !v.starts_with("this."))
			.map(|(v, ops)| (v.to_string(), ops.clone()))
			.collect()
	}

	pub fn types(&self) -> BTreeMap<Vec<String>, TypeRef> {
		let mut res: BTreeMap<Vec<String>, TypeRef> = BTreeMap::new();
		for (n, t) in &self.types {
			res.insert(n.clone(), *t);
		}

		res
	}

	/// Returns the list of all the fields being captures `this.xxx`
	pub fn fields(&self) -> BTreeMap<String, BTreeSet<String>> {
		self
			.vars()
			.iter()
			.filter(|(v, _)| v.starts_with("this."))
			.map(|(v, ops)| (v.to_string(), ops.clone()))
			.collect()
	}

	pub fn fields_without_this(&self) -> IndexMap<String, BTreeSet<String>> {
		self
			.fields()
			.iter()
			.map(|(f, ops)| f.strip_prefix("this.").map(|f| (f.to_string(), ops.clone())))
			.flatten()
			.collect()
	}

	/// For each method, return the list of captured variables and their operations
	pub fn refs(&self) -> BTreeMap<String, BTreeMap<String, BTreeSet<String>>> {
		self.vars.clone()
	}
}

impl Display for ClassCaptures {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		if !self.types.is_empty() {
			write!(f, "Types:\n")?;
			for (name, t) in &self.types {
				write!(f, "  {} = {}\n", name.join("."), t)?;
			}
		}

		if !self.vars.is_empty() {
			write!(f, "Variables:\n")?;
			for (method, vars) in &self.vars {
				write!(f, "  {}():\n", method)?;
				for (var, ops) in vars {
					write!(f, "    {} => [{}]\n", var, ops.iter().join(","))?;
				}
			}
		}

		Result::Ok(())
	}
}

impl<'a> CaptureScanner<'a> {
	pub fn new(
		types: &'a Types,
		method_name: &'a str,
		function_scope: &'a Scope,
		captures: &'a mut ClassCaptures,
	) -> Self {
		Self {
			method_name,
			types,
			captures,
			function_scope,
			method_env: &function_scope.env,
			current_env: &function_scope.env,
			current_index: 0,
		}
	}

	pub fn scan(&mut self) {
		self.visit_scope(self.function_scope);
	}

	fn analyze_expr(&self, node: &'a Expr) -> Vec<Component> {
		let ExprKind::Reference(ref_expr) = &node.kind else {
			return vec![];
		};

		return self.analyze_ref(ref_expr);
	}

	fn analyze_ref(&self, ref_expr: &'a Reference) -> Vec<Component> {
		let env = self.current_env.borrow();

		match ref_expr {
			Reference::Identifier(x) => {
				let lookup = env.as_ref().unwrap().lookup_ext(&x, Some(self.current_index));

				// if the reference is already defined later in this scope, skip it
				if let LookupResult::DefinedLater = lookup {
					return vec![];
				}

				let LookupResult::Found(kind, i) = lookup else {
					panic!("reference to undefined symbol \"{x}\"");
				};

				let var = kind.as_variable().expect("variable");

				// // If the reference isn't a preflight (lifted) variable then skip it
				// if var.phase != Phase::Preflight {
				// 	return vec![];
				// }

				return vec![Component {
					text: x.name.clone(),
					span: x.span.clone(),
					kind: ComponentKind::Member(var, i.env),
				}];
			}
			Reference::InstanceMember {
				object,
				property,
				optional_accessor: _optional_chain,
			} => {
				let obj_type = self.types.get_expr_type(object).unwrap();
				let prop = if let Some(component_kind) = self.to_component_kind(obj_type, property) {
					vec![Component {
						text: property.name.clone(),
						span: property.span.clone(),
						kind: component_kind,
					}]
				} else {
					vec![]
				};

				let obj = self.analyze_expr(&object);
				return [obj, prop].concat();
			}
			Reference::TypeMember { type_, property } => {
				// Get the type we're accessing a member of
				let t = resolve_user_defined_type(type_, &env.as_ref().unwrap(), self.current_index)
					.expect("covered by type checking");

				// check if we are referencing an enum
				if let Type::Enum(_) = &*t {
					return vec![Component {
						text: format!("{type_}"),
						span: type_.span.clone(),
						kind: ComponentKind::ClassType(t),
					}];
				}

				// If the type we're referencing isn't a class then skip it
				let Some(class) = t.as_class() else {
					return vec![];
				};

				// To obtain information about the variable we're referencing (like its type and
				// whether it's reassignable), we look it up in the class's env.
				let (kind, var_info) = class.env.lookup_ext(&property, None).expect("lookup");

				let var = kind.as_variable().expect("variable");

				return vec![
					Component {
						text: format!("{type_}"),
						span: type_.span.clone(),
						kind: ComponentKind::ClassType(t),
					},
					Component {
						text: property.name.clone(),
						span: property.span.clone(),
						kind: ComponentKind::Member(var, var_info.env),
					},
				];
			}
		}
	}

	fn to_component_kind(&self, obj_type: UnsafeRef<Type>, property: &Symbol) -> Option<ComponentKind> {
		fn lookup(env: &SymbolEnv, symbol: &Symbol) -> ComponentKind {
			let (s, i) = env.lookup_ext(&symbol, None).expect("covered by type checking");
			return ComponentKind::Member(s.as_variable().expect("variable"), i.env);
		}

		match &*obj_type {
			Type::Void => unreachable!("cannot reference a member of void"),
			Type::Function(_) => unreachable!("cannot reference a member of a function"),
			Type::Optional(t) => self.to_component_kind(*t, property),
			Type::String
			| Type::Number
			| Type::Duration
			| Type::Boolean
			| Type::Json
			| Type::MutJson
			| Type::Anything
			| Type::Nil
			| Type::Enum(_)
			| Type::Array(_)
			| Type::MutArray(_)
			| Type::Map(_)
			| Type::MutMap(_)
			| Type::Set(_)
			| Type::MutSet(_) => None,

			Type::Class(cls) => Some(lookup(&cls.env, property)),
			Type::Interface(iface) => Some(lookup(&iface.env, property)),
			Type::Struct(st) => Some(lookup(&st.env, property)),
		}
	}

	fn split_phases(&self, parts: &[Component]) -> (Vec<Component>, Vec<Component>) {
		let mut pref = vec![];
		let mut inf = vec![];

		for p in parts {
			match &p.kind {
				ComponentKind::Member(vi, _) => match vi.phase {
					Phase::Preflight => pref.push(p.clone()),
					Phase::Inflight => inf.push(p.clone()),
					Phase::Independent => pref.push(p.clone()),
				},
				ComponentKind::ClassType(_) => {
					pref.push(p.clone());
				}
			}
		}

		(pref, inf)
	}
}

impl<'ast> Visit<'ast> for CaptureScanner<'ast> {
	fn visit_expr_new(
		&mut self,
		node: &'ast Expr,
		class: &'ast TypeAnnotation,
		obj_id: &'ast Option<String>,
		obj_scope: &'ast Option<Box<Expr>>,
		arg_list: &'ast ArgList,
	) {
		// we want to only capture the type annotation in the case of "new X" because
		// other cases of type annotation are actually erased in the javascript code.
		if let TypeAnnotationKind::UserDefined(u) = &class.kind {
			let env = self.current_env.borrow();
			let x = env
				.as_ref()
				.unwrap()
				.lookup_nested(&u.full_path().iter().collect_vec(), Some(self.current_index));

			match x {
				LookupResult::Found(t, _) => self.captures.capture_type(u.full_path_str_vec(), t.as_type().unwrap()),
				LookupResult::NotFound(_) => todo!(),
				LookupResult::DefinedLater => todo!(),
				LookupResult::ExpectedNamespace(_) => todo!(),
			}
		}

		visit::visit_expr_new(self, node, class, obj_id, obj_scope, arg_list);
	}

	fn visit_reference(&mut self, node: &'ast Reference) {
		let parts = self.analyze_ref(node);
		println!("{}", parts.iter().map(|f| f.text.clone()).join("."));
		if parts.is_empty() {
			visit::visit_reference(self, node);
			return;
		}

		// let (pref, inf) = self.split_phases(&parts);

		// determine if this is a reference to a field (through "this.")
		let mut captured_object_index = 0;
		let captured_object = match parts.first() {
			Some(first) => {
				if first.text == "this" {
					captured_object_index = 1;
					parts.get(1)
				} else {
					Some(first)
				}
			}
			None => None,
		};

		let Some(captured_object) = captured_object else {
			return;
		};

		// if we have no parts left, it means we are referencing "this" and we can skip it
		if captured_object_index == parts.len() {
			return;
		}

		// at this point, the first component in "parts" is the object we are capturing and the rest
		// is the qualification of the capture. let's split these up.
		let (captured_object_ref, qualification) = parts.split_at(captured_object_index + 1);
		let captured_object_ref = &captured_object_ref
			.iter()
			.map(|f| f.text.clone())
			.collect_vec()
			.join(".");

		// is this a capture or a local symbol? check if the object's environment is a child or the same
		// as the environment of the method we are currently scanning.
		if let ComponentKind::Member(_, object_env) = captured_object.kind {
			if object_env.is_same(self.method_env.borrow().as_ref().unwrap())
				|| object_env.is_child_of(self.method_env.borrow().as_ref().unwrap())
			{
				return;
			}
		}

		match &captured_object.kind {
			ComponentKind::ClassType(t) => {
				self
					.captures
					.capture_type(captured_object.text.split(".").map(|f| f.to_string()).collect_vec(), *t);
			}
			ComponentKind::Member(variable, _) => {
				// ignore functions that have a js override macro (like `log()`, `assert()`).
				let mut is_function = false;

				if let Type::Function(f) = &*variable.type_ {
					is_function = true;

					if f.js_override.is_some() {
						return;
					}
				}

				// if the varaible's phase is inflight, skip it
				if variable.phase == Phase::Inflight && !is_function {
					// special-handling for inflight methods?
					return;
				}

				// the object we are capturing is a preflight object, so we need to lift it into inflight.
				// we need to make sure that the object is not reassignable and that it has a capturable
				// type.

				// if the variable is reassignable, bail out
				if variable.reassignable {
					report_diagnostic(Diagnostic {
						message: format!("Cannot capture reassignable object '{captured_object}'"),
						span: Some(captured_object.span.clone()),
					});

					return;
				}

				// if this type is not capturable, bail out
				if !variable.type_.is_capturable() {
					report_diagnostic(Diagnostic {
						message: format!(
							"Cannot capture field '{captured_object}' with non-capturable type '{}'",
							variable.type_
						),
						span: Some(captured_object.span.clone()),
					});

					return;
				}

				// okay, so we have a non-reassignable, capturable type.
				// one more special case is collections. we currently only support
				// collections which do not include resources because we cannot
				// qualify the capture.
				if let Some(inner_type) = variable.type_.collection_item_type() {
					if inner_type.is_preflight_class() {
						report_diagnostic(Diagnostic {
							message: format!(
								"Capturing collection of preflight classes is not supported yet (type is '{}')",
								variable.type_,
							),
							span: Some(captured_object.span.clone()),
						});

						return;
					}
				}

				let op = if !qualification.is_empty() {
					Some(qualification.iter().map(|f| f.text.clone()).collect_vec().join("."))
				} else {
					None
				};

				self.captures.capture_var(self.method_name, captured_object_ref, op);

				return;
			}
		}

		visit::visit_reference(self, node);
	}

	fn visit_scope(&mut self, node: &'ast Scope) {
		let backup_env = self.current_env;
		self.current_env = &node.env;
		visit::visit_scope(self, node);
		self.current_env = backup_env;
	}

	fn visit_stmt(&mut self, node: &'ast Stmt) {
		self.current_index = node.idx;
		visit::visit_stmt(self, node);
	}
}

#[cfg(test)]
mod tests;
