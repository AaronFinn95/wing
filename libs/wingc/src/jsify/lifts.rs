use itertools::Itertools;
use std::collections::{BTreeMap, BTreeSet};

#[derive(Default)]
pub struct Lifts {
	lifts: BTreeMap<String, LiftedObject>,
}

#[derive(Clone)]
pub struct LiftedObject {
	pub method_name: String,
	pub is_field: bool,
	pub inflight_symbol: String,
	pub preflight_code: String,
	pub operations: BTreeSet<String>,
}

impl LiftedObject {
	pub fn render_ops_js_array(&self) -> String {
		let ops = self
			.operations
			.iter()
			.map(|f| format!("\"{}\"", f))
			.collect_vec()
			.join(", ");
		format!("[{}]", ops)
	}
}



impl Lifts {
	pub fn add(&mut self, method_name: String, is_field: bool, preflight_code: String, op: Option<String>) -> String {
		let prefix = if is_field { "this." } else { "" };

		let slug = replace_non_alphanumeric(&preflight_code);
		let inflight_symbol = format!("{prefix}${}", slug);

		let key = format!("{}.{}", method_name, inflight_symbol);

		let x = self.lifts.entry(key).or_insert(LiftedObject {
			method_name: method_name.clone(),
			inflight_symbol: inflight_symbol.clone(),
			preflight_code: preflight_code.clone(),
			is_field,
			operations: BTreeSet::default(),
		});

		if let Some(op) = op {
			x.operations.insert(op);
		}

		inflight_symbol
	}

	pub fn per_method(&self) -> BTreeMap<String, Vec<LiftedObject>> {
		let mut result: BTreeMap<String, Vec<LiftedObject>> = BTreeMap::new();

		for lift in &self.all() {
			result.entry(lift.method_name.clone()).or_default().push(lift.clone());
		}

		result.clone()
	}

	/// Returns all lifted objects, deduplicated.
	pub fn all(&self) -> Vec<LiftedObject> {
		let mut map = BTreeMap::<String, LiftedObject>::new();

		for l in self.lifts.values() {
			let o = map.entry(l.inflight_symbol.clone()).or_insert(LiftedObject {
				method_name: l.method_name.clone(),
				is_field: l.is_field,
				inflight_symbol: l.inflight_symbol.clone(),
				preflight_code: l.preflight_code.clone(),
				operations: BTreeSet::default(),
			});

			o.operations.extend(l.operations.clone());
		}

		let mut result: Vec<LiftedObject> = vec![];
		for x in map.values() {
			result.push(x.clone());
		}

		result
	}
}

fn replace_non_alphanumeric(s: &str) -> String {
	let mut result = String::new();

	for c in s.chars() {
		if c.is_alphanumeric() {
			result.push(c);
		} else {
			result.push('_');
		}
	}

	result
}
