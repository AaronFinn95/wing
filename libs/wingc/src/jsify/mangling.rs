use std::collections::{BTreeMap, BTreeSet};

use itertools::Itertools;

use crate::jsify::codemaker::CodeMaker;

#[derive(Clone)]
pub enum JavaScriptObject {
	Value(String),
	Object(BTreeMap<String, JavaScriptObject>),
}

pub struct MangleDemangle {
	pub keys: Vec<String>,
	pub objects: BTreeMap<String, JavaScriptObject>,
	pub javascript: String,
}

pub fn mangle_captures(captures: BTreeSet<String>) -> MangleDemangle {
	let mut result = MangleDemangle {
		keys: vec![],
		objects: BTreeMap::new(),
		javascript: String::new(),
	};

	fn add_to_object(path: Vec<&str>, parent: &mut BTreeMap<String, JavaScriptObject>, value: String) {
		let Some(first) = path.first() else {
			return; // no more components in the path
		};

		// if this is the last element, insert it as a value to the parent
		if path.len() == 1 {
			parent.insert(first.to_string(), JavaScriptObject::Value(value));
			return;
		}

		// otherwise, this is an object, so we need to create it if it doesn't exist
		let child = parent
			.entry(first.to_string())
			.or_insert(JavaScriptObject::Object(BTreeMap::new()));

		// if we already captured this key as a value, then we can skip the nested objects
		// because the entire object is already captured.
		let JavaScriptObject::Object(child) = child else {
			return;
		};

		// and recurse
		add_to_object(path[1..].to_vec(), child, value);
	}

	// we sort in order to be able to skip nested objects if the parent has already been captured
	for v in captures.iter().sorted() {
		let key = mangle(&v);
		result.keys.push(key.clone());
		add_to_object(v.split(".").collect_vec(), &mut result.objects, key.clone());
	}

	// format the objects as javascript objects
	for (k, v) in &result.objects {
		if let JavaScriptObject::Object(o) = v {
			let mut code = CodeMaker::default();
			code.open(format!("const {} = {{", k));
			format_js_obj(o, &mut code);
			code.close("};");

			result.javascript.push_str(&code.to_string());
		}
	}

	result
}

#[derive(PartialEq, Debug)]

pub struct LiftArgs {
	pub key: String,
	pub field: String,
	pub ops: String,
}

/// Creates a map from object names to a JavaScript array of operations that are performed on that
/// object. This is used when generating calls to `_lift(obj, ops)`.
pub fn make_lift_args(vars: BTreeMap<String, BTreeSet<String>>) -> Vec<LiftArgs> {
	let mut result = vec![];

	for (v, ops) in vars {
		let arr = ops.iter().map(|f| format!("\"{f}\"").clone()).collect_vec().join(", ");
		result.push(LiftArgs {
			key: mangle(&v),
			field: v.clone(),
			ops: format!("[{arr}]"),
		});
	}

	return result;
}

pub fn mangle(s: &str) -> String {
	s.replace(".", "_")
}

pub fn format_js_obj(map: &BTreeMap<String, JavaScriptObject>, code: &mut CodeMaker) {
	for (k, v) in map {
		match v {
			JavaScriptObject::Value(v) => {
				code.line(&format!("{}: {},", k, v));
			}
			JavaScriptObject::Object(o) => {
				code.open(&format!("{}: {{", k));
				format_js_obj(o, code);
				code.close("},");
			}
		}
	}
}

#[cfg(test)]
mod tests;
