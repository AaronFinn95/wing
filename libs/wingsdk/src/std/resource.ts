import { Construct, IConstruct } from "constructs";
import { Datetime } from "./datetime";
import { Duration } from "./duration";
import { App, Bindings, Connections } from "../core";

/**
 * A resource that can run inflight code.
 * @skipDocs
 */
export interface IInflightHost extends IConstruct {}

/**
 * A resource with inflight operations.
 */
export interface IInflightConstruct extends IConstruct {
  /**
   * Return a code snippet that can be used to reference this resource inflight.
   *
   * Note this code snippet may by async code, so it's unsafe to run it in a
   * constructor or other sync context.
   *
   * @internal
   */
  _toInflight(): string;
}

/**
 * This interface contains documentation for methods that the Wing compiler generates but may
 * not be needed by all constructs.
 */
// @ts-ignore allow unused declaration
interface IExtraInflightOperations {
  /**
   * Binds the resource to the host so that it can be used by inflight code.
   *
   * If a construct implements this method, it will not be called no more than
   * once with a given `host`.
   *
   * If a construct doesn't implement this method, it just means that none of its
   * inflight operations add any requirements on the host (like IAM permissions or
   * environment variables).
   *
   * This method shouldn't call `bind` on any other constructs besides its parent `super`.
   */
  bind(host: IInflightHost, ops: string[]): void;

  /**
   * Return a list of supported inflight operations supported by the resource.
   *
   * If a construct doesn't implement this method, it's assumed that the construct
   * doesn't support any inflight operations.
   */
  _getInflightOps(): string[];

  /**
   * Register any relationships between
   * the class and any transitively referenced classes. For example, if calling
   * the inflight method "foo" on a class requires calling the inflight methods
   * "bar" and "baz" on another class, then the compiler might generate:
   *
   * ```
   * _registerBind(host, ops) {
   *   if (ops.includes("foo")) {
   *     this._bar._registerBind(host, ["bar", "baz"]);
   *   }
   * }
   * ```
   *
   * The method's implementation shouldn't contain any actual binding logic
   * (like adding IAM permissions or environment variables to the host).
   *
   * If the method isn't implemented, it can be assumed that the class doesn't
   * reference any other classes.
   */
  _registerBind(host: IInflightHost, ops: string[]): void;

  /**
   * (static method) Register that the resource type needs to be bound to the host for the given
   * operations. A type being bound to a host means that that type's static members
   * will be bound to the host.
   */
  _registerTypeBind(host: IInflightHost, ops: string[]): void;
}

/**
 * Shared behavior between all Wing SDK resources.
 * @skipDocs
 */
export class Resource {
  /**
   * Register a binding between an object (either data or class or instance) and a host.
   * All of the registered bindings will be aggregated, so that when synthesis is
   * performed, all of the necessary `bind()` calls will be invoked.
   *
   * - Primitives, Duration, and Datetime objects are ignored.
   * - Arrays, sets and maps and structs (Objects) are recursively bound.
   * - Class instances are bound to the host by calling their bind() method.
   * - (TODO) Classes and their static methods are bound to their host by calling their bindType() method.
   *
   * @param obj The object to bind.
   * @param host The host to bind to
   * @param ops The set of operations that may access the object
   *
   * @internal
   */
  public static _registerBindObject(
    obj: any,
    host: IInflightHost,
    ops: string[] = []
  ): void {
    const tokens = App.of(host)._tokens;
    if (tokens.isToken(obj)) {
      return tokens.bindValue(host, obj);
    }

    switch (typeof obj) {
      case "string":
      case "boolean":
      case "number":
      case "undefined":
        return;

      case "object":
        if (Array.isArray(obj)) {
          obj.forEach((item) => Resource._registerBindObject(item, host));
          return;
        }

        if (obj instanceof Duration) {
          return;
        }

        if (obj instanceof Datetime) {
          return;
        }

        if (obj instanceof Set) {
          return Array.from(obj).forEach((item) =>
            Resource._registerBindObject(item, host)
          );
        }

        if (obj instanceof Map) {
          Array.from(obj.values()).forEach((item) =>
            Resource._registerBindObject(item, host)
          );
          return;
        }

        // if the object is a resource (i.e. has a "bind" method"), register a binding between it and the host.
        if (Construct.isConstruct(obj)) {
          // Explicitly register the resource's `$inflight_init` op, which is a special op that can be used to makes sure
          // the host can instantiate a client for this resource.
          let extendedOps = [...ops, "$inflight_init"];

          const bindings = Bindings.of(obj);

          // For each operation, check if the host supports it. If it does, register the binding.
          const supportedOps = [
            ...((obj as any)._getInflightOps?.() ?? []),
            "$inflight_init",
          ];

          for (const op of extendedOps) {
            if (!supportedOps.includes(op)) {
              throw new Error(
                `Resource ${obj.node.path} does not support inflight operation ${op} (requested by ${host.node.path})`
              );
            }

            if (bindings.has(host, op)) {
              // this resource is already bound to the host for this operation
              continue;
            }

            // add the operation to the set of operations for the host so that we can avoid
            // infinite recursion.
            bindings.add(host, op);

            if (typeof (obj as any)._registerBind === "function") {
              (obj as any)._registerBind?.(host, [op]);
            }

            // add connection metadata
            Connections.of(obj).add({
              source: host,
              target: obj,
              name: op.endsWith("()") ? op : `${op}()`,
            });
          }
          return;
        }

        // structs are just plain objects
        if (obj.constructor.name === "Object") {
          Object.values(obj).forEach((item) =>
            Resource._registerBindObject(item, host, ops)
          );
          return;
        }
        break;

      case "function":
        // If the object is actually a resource type, call the type's _registerTypeBind static method
        if (typeof (obj as any)._registerTypeBind === "function") {
          obj._registerTypeBind(host, ops);
          return;
        }
        break;
    }

    throw new Error(
      `unable to serialize immutable data object of type ${obj.constructor?.name}`
    );
  }
}
