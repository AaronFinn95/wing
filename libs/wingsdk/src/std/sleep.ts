import { Duration } from "./duration";

export const sleep = async (duration: Duration): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve, duration.milliseconds));
