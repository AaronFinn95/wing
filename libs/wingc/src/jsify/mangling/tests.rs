use std::{collections::BTreeSet, fmt::Display};

use insta::assert_snapshot;

use crate::jsify::mangle_captures;

use super::MangleDemangle;

fn as_set(x: &[&str]) -> BTreeSet<String> {
	x.iter().map(|s| s.to_string()).collect::<BTreeSet<_>>()
}

#[test]
fn simple_objects() {
	let result = mangle_captures(as_set(&["a", "hello", "MyClass"]));
	assert_snapshot!(result.to_string());
}

#[test]
fn nested_objects() {
	let result = mangle_captures(as_set(&["a.b.c", "a.b.d", "a.b.e"]));
	assert_snapshot!(result.to_string());
}

#[test]
fn mixed() {
	let result = mangle_captures(as_set(&[
		"a.b.c",
		"a.b.d",
		"a.b.e",
		"hello.world",
		"b.myBucket",
		"hello",
		"MyClass",
		"std.Json",
		"std.Number",
	]));
	assert_snapshot!(result.to_string());
}

impl Display for MangleDemangle {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		writeln!(f, "keys: {}", self.keys.join(","))?;
		writeln!(f, "\n{}", self.javascript)
	}
}
