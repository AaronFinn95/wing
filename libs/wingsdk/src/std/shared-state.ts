import { Construct } from "constructs";
import { IResource, Resource } from "./resource";
import { fqnForType } from "../constants";
import { App } from "../core";

/**
 * Global identifier for `SharedState`.
 */
export const SHARED_STATE_FQN = fqnForType("std.SharedState");

/**
 * A container for a value that can be shared and mutated across inflight consumers.
 * Only supported in the `sim` compilation target.
 *
 * @inflight `@winglang/sdk.std.ISharedStateClient`
 */
export abstract class SharedState extends Resource implements IResource {
  /**
   * Create a new SharedState instance.
   * @internal
   */
  public static _newSharedState(
    scope: Construct,
    id: string,
    initial: any
  ): SharedState {
    return App.of(scope).newAbstract(SHARED_STATE_FQN, scope, id, initial);
  }

  constructor(scope: Construct, id: string, initial: any) {
    super(scope, id);

    initial;

    this.display.title = "SharedState";
    this.display.description =
      "A container that can share a value across inflight consumers.";

    this._addInflightOps(SharedStateMethods.GET, SharedStateMethods.SET);
  }
}

/**
 * Inflight interface for `SharedState`.
 */
export interface ISharedStateClient {
  /**
   * Get the inner value.
   * @inflight
   */
  get(): Promise<any>;

  /**
   * Replace the inner value with a new value.
   * @inflight
   */
  set(value: any): Promise<void>;
}

/**
 * List of inflight operations available for `SharedState`.
 * @internal
 */
export enum SharedStateMethods {
  /** `SharedState.get` */
  GET = "get",
  /** `SharedState.set` */
  SET = "set",
}
