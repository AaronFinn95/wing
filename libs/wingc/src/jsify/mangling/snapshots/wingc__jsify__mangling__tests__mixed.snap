---
source: libs/wingc/src/jsify/mangling/tests.rs
expression: result.to_string()
---
keys: MyClass,a_b_c,a_b_d,a_b_e,b_myBucket,hello,hello_world,std_Json,std_Number

const a = {
  b: {
    c: a_b_c,
    d: a_b_d,
    e: a_b_e,
  },
};
const b = {
  myBucket: b_myBucket,
};
const std = {
  Json: std_Json,
  Number: std_Number,
};


